plugins/auth-backend/src/providers/microsoft/provider.ts

```diff
@@ -165,37 +192,102 @@ export class MicrosoftAuthProvider implements OAuthHandlers {
           resolve(photoURL);
         })
         .catch(error => {
-          console.log(
+          this.logger.warn(
             `Could not retrieve user profile photo from Microsoft Graph API: ${error}`,
           );
           // User profile photo is optional, ignore errors and resolve undefined
           resolve(undefined);
         });
     });
   }
+}
 
-  private async populateIdentity(
-    response: OAuthResponse,
-  ): Promise<OAuthResponse> {
-    const { profile } = response;
+export const microsoftEmailSignInResolver: SignInResolver<OAuthResult> = async (
+  info,
+  ctx,
+) => {
+  const { profile } = info;
 
-    if (!profile.email) {
-      throw new Error('Microsoft profile contained no email');
-    }
+  if (!profile.email) {
+    throw new Error('Microsoft profile contained no email');
+  }
+
+  const entity = await ctx.catalogIdentityClient.findUser({
+    annotations: {
+      'microsoft.com/email': profile.email,
+    },
+  });
+
+  const claims = getEntityClaims(entity);
+  const token = await ctx.tokenIssuer.issueToken({ claims });
 
-    // Like Google implementation, setting this to local part of email for now
-    const id = profile.email.split('@')[0];
+  return { id: entity.metadata.name, entity, token };
+};
+
+export const microsoftDefaultSignInResolver: SignInResolver<OAuthResult> = async (
+  info,
+  ctx,
+) => {
+  const { profile } = info;
 
-    return { ...response, backstageIdentity: { id } };
+  if (!profile.email) {
+    throw new Error('Profile contained no email');
   }
-}
 
-export type MicrosoftProviderOptions = {};
+  let userId: string;
+  try {
+    const entity = await ctx.catalogIdentityClient.findUser({
```

Comment:
The default sign-in resolver is just here for backwards compatibility and should mirror the existing logic, so we should skip the catalog query and just do the email check and split from the old `populateIdentity`